src/sync/mod.rs

```diff
@@ -0,0 +1,340 @@
+        };
+
+        log::info!(
+            "Found a new unfinished anchoring transaction proposal for height: {}",
+            block_height
+        );
+
+        let mut signer = p2wsh::InputSigner::new(redeem_script);
+        let sign_input_messages = inputs
+            .iter()
+            .enumerate()
+            .map(|(index, proposal_input)| {
+                let signature = signer.sign_input(
+                    TxInRef::new(proposal.as_ref(), index),
+                    proposal_input.as_ref(),
+                    &(keypair.1).0.key,
+                )?;
+
+                Ok(SignInput {
+                    input: index as u32,
+                    input_signature: signature.into(),
+                    txid: proposal.id(),
+                })
+            })
+            .collect::<Result<Vec<_>, failure::Error>>()
+            .map_err(ChainUpdateError::Internal)?;
+        // Send sign input transactions to the Exonum node.
+        for sign_input in sign_input_messages {
+            self.api_client
+                .sign_input(sign_input)
+                .wait()
+                .map_err(ChainUpdateError::Client)?;
+        }
+        Ok(())
+    }
+
+    fn find_private_key(
+        &self,
+        anchoring_keys: impl IntoIterator<Item = btc::PublicKey>,
+    ) -> Option<(btc::PublicKey, btc::PrivateKey)> {
+        anchoring_keys.into_iter().find_map(|public_key| {
+            self.key_pool
+                .get(&public_key)
+                .cloned()
+                .map(|private_key| (public_key, private_key))
+        })
+    }
+}
+
+/// Errors that occur when updating the sync with Bitcoin task.
+#[derive(Debug)]
+pub enum SyncWithBitcoinError<C: Display, R: Display> {
+    /// Error occurred in the private API client.
+    Client(C),
+    /// Error occurred in the Bitcoin relay.
+    Relay(R),
+    /// Internal error.
+    Internal(failure::Error),
+    /// Initial funding transaction is unconfirmed.
+    UnconfirmedFundingTransaction(btc::Sha256d),
+}
+
+/// Pushes anchoring transactions to the Bitcoin blockchain.
+#[derive(Debug)]
+pub struct SyncWithBitcoinTask<T, R>
+where
+    T: PrivateApi + 'static,
+    R: BitcoinRelay + 'static,
+{
+    btc_relay: R,
+    api_client: T,
+}
+
+impl<T, R> SyncWithBitcoinTask<T, R>
+where
+    T: PrivateApi + 'static,
+    R: BitcoinRelay + 'static,
+    T::Error: Display,
+    R::Error: Display,
+{
+    /// Creates a new sync with Bitcoin task instance.
+    pub fn new(btc_relay: R, api_client: T) -> Self {
+        Self {
+            api_client,
+            btc_relay,
+        }
+    }
+
+    /// Performs one attempt to send the first uncommitted anchoring transaction into the Bitcoin network, if any.
+    /// sign an anchoring proposal, if any. Returns an index of the last committed transaction.
+    pub fn process(
+        &self,
+        latest_committed_tx_index: Option<u64>,
+    ) -> Result<Option<u64>, SyncWithBitcoinError<T::Error, R::Error>> {
```

Comment:
This signature is really scary =_=

Unintantionall (I guess) pun in the type `T::error` describes it quite well.